storage2/src/state/transaction.rs

```diff
@@ -5,22 +5,16 @@ use super::StateWrite;
 /// Represents a transactional set of changes to a `State` fork,
 /// implemented as a RYW cache over a `State`.
 pub struct Transaction<'a> {
-    // TODO determine which fields to include (#1490)
-    // cache: HashMap<jmt::KeyHash, jmt::OwnedValue>,
-    // unwritten_changes: Vec<(jmt::KeyHash, jmt::OwnedValue)>,
+    pub(crate) unwritten_changes: Vec<(String, Option<Vec<u8>>)>,
```

Comment:
I think this should be a `HashMap<String, Option<Vec<u8>>>`, not a `Vec`. That way if you overwrite a key multiple times, or delete it after inserting it in the same unwritten set, you don't have extra memory used to represent those useless operations, and they don't have to be processed by `put_value_set`. This works now that `put_value_set` takes any `IntoIterator<Item = (KeyHash, Option<OwnedValue>)>` as an argument, rather than just a `Vec`.

Reply:
Rather than a `HashMap`, it'd be better to use a `BTreeMap`, so that we'll be able to iterate over keys in order.